Add document describing how to disable tests

BruceForstall opened the pull request:
Update libraries "filtering tests" doc with information and examples for
using `SkipOnCoreClrAttribute`

BruceForstall committed:
Add document describing how to disable tests

Update libraries "filtering tests" doc with information and examples for
using `SkipOnCoreClrAttribute`

BruceForstall commented:
@dotnet/runtime-infrastructure @danmoseley I wrote a document on how to disable tests, and added to the libraries test filtering document information on using `SkipOnCoreClr` (as well as adding various examples). PTAL.

trylek reviewed (approved):
Beautiful, thanks Bruce for putting this together!

jkoritzinsky commented:
You should note that all of the attributes used in the libraries tests are also supported in the runtime tests when using xunit attributes.

jkotas commented on docs/workflow/testing/libraries/filtering-tests.md:
````diff
@@ -176,3 +234,67 @@ dotnet build <csproj_file> /t:Test /p:TargetOS=windows
 dotnet build <csproj_file> /t:Test /p:TargetOS=OSX /p:WithCategories="OuterLoop;failing""
 ```
+
+## SkipOnCoreClrAttribute
````
We also `SkipOnMonoAttribute`. 

On average, there are more tests disabled on Mono than on CoreCLR, so `SkipOnMonoAttribute` is more common.

jkotas commented on docs/workflow/testing/libraries/filtering-tests.md:
````diff
@@ -176,3 +234,67 @@ dotnet build <csproj_file> /t:Test /p:TargetOS=windows
 ```
+
+## SkipOnCoreClrAttribute
+This attribute is used to disable a test under specific conditions, only when run with CoreCLR (it doesn't affect tests run with Mono). Typically, this is when there is a failure in a particular test run configuration, such as under GCStress or JitStress.
````
I believe that `SkipOn...` attributes are meant to be used for permanent disable, without an ActiveIssue.

danmoseley replied:
ideally, but there are some filters possible on them that aren't possible on ActiveIssue.

danmoseley commented on docs/workflow/testing/libraries/filtering-tests.md:
````diff
@@ -176,3 +234,67 @@ dotnet build <csproj_file> /t:Test /p:TargetOS=windows
+[SkipOnCoreClr("Long running tests: https://github.com/dotnet/runtime/issues/11980", TestPlatforms.Linux)
+```
+
+**Disable for specific test mode:**
````
Is there a way to disable a test for any "slow" mode? that might be a debug runtime, or some stress mode. Not sure whether that's a scenario.

danmoseley replied:
Ah yes you can have multiple such attributes

        [SkipOnCoreClr("https://github.com/dotnet/runtime/issues/67886", ~RuntimeConfiguration.Release)]
        [SkipOnCoreClr("https://github.com/dotnet/runtime/issues/67886", RuntimeTestModes.JitMinOpts)]

danmoseley commented on docs/workflow/ci/disabling-tests.md:
```diff
@@ -0,0 +1,71 @@
+
+## Runtime or libraries?
+
+There are two main sets of tests in the repo: runtime tests in the src/tests tree, and libraries tests which
```
are there any CLR specific or Mono specific test trees now, or are they all just subsets of the runtime tests in src/tests?

BruceForstall replied:
IIUC, it is the same test tree, but I don't know if Mono has something in addition.

danmoseley commented on docs/workflow/testing/libraries/filtering-tests.md:
````diff
@@ -1,8 +1,18 @@
+
+## OuterLoopAttribute
 
 ```cs
````
another outerloop reason is that the test is disruptive, eg., it pops up a dialog.

we have been de-facto using outerloop for tests that are a bit less reliable, although that practice is dubious now that we are driving outerloop clean.

danmoseley commented on docs/workflow/testing/libraries/filtering-tests.md:
````diff
@@ -176,3 +234,67 @@ dotnet build <csproj_file> /t:Test /p:TargetOS=windows
+SkipOnCoreClr(string reason, TestPlatforms testPlatforms, RuntimeConfiguration runtimeConfigurations)
+SkipOnCoreClr(string reason, TestPlatforms testPlatforms, RuntimeTestModes testMode)
+SkipOnCoreClr(string reason, TestPlatforms testPlatforms, RuntimeConfiguration runtimeConfigurations, RuntimeTestModes testModes)
+```
````
it's probably out of scope, but nowhere better to put it -- `[Collection(nameof(DisableParallelization))]` anywhere. We put it on test classes to indicate that none of the tests in that class (which as usual run serially with respect to each other) may run concurrently with tests in another class. We use this for tests that use a lot of disk space or memory, or dominate all the cores, such that they are likely to disrupt any tests that run concurrently.

danmoseley reviewed (approved):
thanks for creating these

BruceForstall committed:
Update for feedback

Added SkipOnMonoAttribute, CollectionAttribute, an example
of using multiple SkipOnCoreClrAttribute, and how the xunit
attributes for test disabling apply to the "converted"
src/tests tree.

ViktorHofer reviewed (commented):
LGTM, thanks.

Ideally we would move the content from "filtering-tests.md" into the XunitExtensions arcade project as a README.md, include that file into the package and link to it from here.
